### ISSUE TITLE
Reproducibility Issues with Reported Accuracy

### ISSUE THREAD

**Issue by miralys1** [NVlabs/MambaVision#57]
Hi,

Thank you for your work on MambaVision! I have been trying to reproduce the reported results, specifically for the model MambaVision-T and ran into some problems. 

Here is the summary:
- I ran 16 training runs, using up to 4 GPUs. To match the global batch size, I used a larger per-GPU batch size.
- I experimented with different seeds, but even with the same seed, I observed fluctuations of ~0.1-0.2 percentage points in accuracy.
- My highest achieved accuracy is 82.21%, whereas the reported result is 82.3%.
- When validating using the provided model checkpoint with validate.sh, I get an accuracy of 82.244%, which does not round to 82.3%.

My environment:
- Python 3.10.12 
- torch==2.5.1
- timm==1.0.14
- einops==0.8.0
- transformers==4.48.1
- causal-conv1d @ file:///causal-conv1d (using the newest commit as of this post: 82867a9)
- mamba-ssm @ file:///mamba (using the newest commit as of this post: 0cce0fa)

Could you clarify if there are any additional details regarding the training setup or hyperparameters that might explain these discrepancies? Also, was any additional post-processing or averaging applied to obtain the reported accuracy?

**Comment by ahatamiz** [comment]
Hi @miralys1 

Thanks for your interest in our work. 

The training setup exactly follows our config [file](https://github.com/NVlabs/MambaVision/blob/main/mambavision/configs/mambavision_tiny_1k.yaml). We use a global batch size of 4096. As a result, you need to account for any changes with respect to per gpu batch sizes depending your hardware setup. We initially used 4 nodes for training with the local batch size of 128. 

The environment setup follows our docker [file](https://github.com/NVlabs/MambaVision/blob/main/Dockerfile). 

There are no additional post-processing applied to obtain the reported accuracy. I just downloaded the model weights and reproduced the results once again from scratch in my own system. Please see the logs below:

```
Validating in float32. AMP not enabled.
Loaded state_dict from checkpoint './models/mamba_vision/mambavision_tiny_1k.pth.tar'
Model mamba_vision_T created, param count: 31794248
Data processing configuration for current model + dataset:
        input_size: (3, 224, 224)
        interpolation: bicubic
        mean: (0.485, 0.456, 0.406)
        std: (0.229, 0.224, 0.225)
        crop_pct: 1.0
        crop_mode: center
Test: [   0/391]  Time: 1.624s (1.624s,   78.80/s)  Loss:  0.5208 (0.5208)  Acc@1:  94.531 ( 94.531)  Acc@5:  98.438 ( 98.438)
Test: [  10/391]  Time: 0.038s (0.409s,  312.87/s)  Loss:  1.1768 (0.6252)  Acc@1:  76.562 ( 92.330)  Acc@5:  96.875 ( 98.366)
Test: [  20/391]  Time: 0.583s (0.439s,  291.90/s)  Loss:  0.7822 (0.7693)  Acc@1:  87.500 ( 87.388)  Acc@5:  97.656 ( 97.954)
Test: [  30/391]  Time: 0.038s (0.406s,  315.32/s)  Loss:  1.1176 (0.8752)  Acc@1:  85.938 ( 84.526)  Acc@5:  94.531 ( 97.354)
Test: [  40/391]  Time: 0.958s (0.399s,  320.97/s)  Loss:  0.6889 (0.8090)  Acc@1:  92.188 ( 86.528)  Acc@5:  95.312 ( 97.523)
Test: [  50/391]  Time: 0.038s (0.377s,  339.32/s)  Loss:  0.4879 (0.8042)  Acc@1:  95.312 ( 86.857)  Acc@5:  99.219 ( 97.518)
Test: [  60/391]  Time: 0.742s (0.384s,  333.66/s)  Loss:  0.9504 (0.7709)  Acc@1:  86.719 ( 87.756)  Acc@5:  94.531 ( 97.682)
Test: [  70/391]  Time: 0.053s (0.365s,  350.23/s)  Loss:  0.9047 (0.7994)  Acc@1:  81.250 ( 86.939)  Acc@5:  99.219 ( 97.568)
Test: [  80/391]  Time: 1.320s (0.379s,  337.54/s)  Loss:  0.9035 (0.8149)  Acc@1:  83.594 ( 86.507)  Acc@5:  96.875 ( 97.512)
Test: [  90/391]  Time: 0.038s (0.369s,  346.63/s)  Loss:  1.0249 (0.8170)  Acc@1:  73.438 ( 86.564)  Acc@5:  97.656 ( 97.433)
Test: [ 100/391]  Time: 0.997s (0.369s,  346.49/s)  Loss:  0.6637 (0.8194)  Acc@1:  94.531 ( 86.324)  Acc@5:  96.875 ( 97.556)
Test: [ 110/391]  Time: 0.038s (0.361s,  354.09/s)  Loss:  1.2035 (0.8286)  Acc@1:  63.281 ( 85.938)  Acc@5:  99.219 ( 97.600)
Test: [ 120/391]  Time: 0.918s (0.365s,  351.02/s)  Loss:  0.6973 (0.8259)  Acc@1:  90.625 ( 86.151)  Acc@5:  97.656 ( 97.630)
Test: [ 130/391]  Time: 0.038s (0.369s,  347.01/s)  Loss:  0.4916 (0.8191)  Acc@1:  96.875 ( 86.397)  Acc@5:  99.219 ( 97.609)
Test: [ 140/391]  Time: 0.038s (0.367s,  348.55/s)  Loss:  1.1913 (0.8191)  Acc@1:  75.000 ( 86.264)  Acc@5:  99.219 ( 97.689)
Test: [ 150/391]  Time: 0.038s (0.373s,  343.38/s)  Loss:  0.8972 (0.8249)  Acc@1:  78.125 ( 86.170)  Acc@5:  98.438 ( 97.615)
Test: [ 160/391]  Time: 0.039s (0.365s,  351.16/s)  Loss:  0.8832 (0.8263)  Acc@1:  85.938 ( 86.156)  Acc@5:  96.875 ( 97.583)
Test: [ 170/391]  Time: 0.038s (0.367s,  349.08/s)  Loss:  1.5230 (0.8429)  Acc@1:  66.406 ( 85.650)  Acc@5:  89.062 ( 97.428)
Test: [ 180/391]  Time: 0.038s (0.365s,  351.08/s)  Loss:  1.5031 (0.8557)  Acc@1:  63.281 ( 85.342)  Acc@5:  95.312 ( 97.272)
Test: [ 190/391]  Time: 0.038s (0.372s,  344.04/s)  Loss:  1.4004 (0.8734)  Acc@1:  69.531 ( 84.833)  Acc@5:  94.531 ( 97.129)
Test: [ 200/391]  Time: 0.038s (0.367s,  348.55/s)  Loss:  0.7788 (0.8900)  Acc@1:  88.281 ( 84.507)  Acc@5:  98.438 ( 96.949)
Test: [ 210/391]  Time: 0.038s (0.367s,  348.72/s)  Loss:  1.0119 (0.9015)  Acc@1:  81.250 ( 84.171)  Acc@5:  96.875 ( 96.879)
Test: [ 220/391]  Time: 0.038s (0.364s,  351.99/s)  Loss:  0.6047 (0.9016)  Acc@1:  91.406 ( 84.188)  Acc@5:  98.438 ( 96.847)
Test: [ 230/391]  Time: 0.038s (0.365s,  350.28/s)  Loss:  1.3498 (0.9062)  Acc@1:  73.438 ( 84.145)  Acc@5:  92.969 ( 96.784)
Test: [ 240/391]  Time: 0.038s (0.363s,  352.72/s)  Loss:  0.9668 (0.9066)  Acc@1:  86.719 ( 84.197)  Acc@5:  93.750 ( 96.726)
Test: [ 250/391]  Time: 0.038s (0.364s,  351.17/s)  Loss:  0.6347 (0.9198)  Acc@1:  92.188 ( 83.721)  Acc@5:  98.438 ( 96.595)
Test: [ 260/391]  Time: 0.038s (0.360s,  355.57/s)  Loss:  0.9309 (0.9283)  Acc@1:  80.469 ( 83.435)  Acc@5:  96.875 ( 96.534)
Test: [ 270/391]  Time: 0.038s (0.360s,  355.07/s)  Loss:  1.4353 (0.9334)  Acc@1:  69.531 ( 83.323)  Acc@5:  92.188 ( 96.460)
Test: [ 280/391]  Time: 0.038s (0.358s,  357.57/s)  Loss:  1.0734 (0.9359)  Acc@1:  76.562 ( 83.246)  Acc@5:  96.875 ( 96.444)
Test: [ 290/391]  Time: 0.038s (0.360s,  355.71/s)  Loss:  1.4248 (0.9410)  Acc@1:  53.125 ( 83.102)  Acc@5:  96.875 ( 96.370)
Test: [ 300/391]  Time: 0.038s (0.359s,  356.97/s)  Loss:  0.9001 (0.9439)  Acc@1:  83.594 ( 83.062)  Acc@5:  96.094 ( 96.333)
Test: [ 310/391]  Time: 0.039s (0.357s,  358.40/s)  Loss:  0.9354 (0.9474)  Acc@1:  84.375 ( 82.993)  Acc@5:  96.094 ( 96.277)
Test: [ 320/391]  Time: 0.038s (0.358s,  357.38/s)  Loss:  0.6966 (0.9522)  Acc@1:  91.406 ( 82.842)  Acc@5:  97.656 ( 96.235)
Test: [ 330/391]  Time: 0.038s (0.357s,  358.57/s)  Loss:  1.3293 (0.9609)  Acc@1:  67.188 ( 82.574)  Acc@5:  93.750 ( 96.155)
Test: [ 340/391]  Time: 0.038s (0.358s,  357.67/s)  Loss:  0.8611 (0.9639)  Acc@1:  84.375 ( 82.494)  Acc@5:  98.438 ( 96.142)
Test: [ 350/391]  Time: 0.038s (0.355s,  360.57/s)  Loss:  0.9366 (0.9653)  Acc@1:  81.250 ( 82.414)  Acc@5:  96.875 ( 96.140)
Test: [ 360/391]  Time: 0.038s (0.358s,  357.45/s)  Loss:  1.2927 (0.9720)  Acc@1:  74.219 ( 82.217)  Acc@5:  98.438 ( 96.105)
Test: [ 370/391]  Time: 0.038s (0.355s,  360.52/s)  Loss:  1.0920 (0.9693)  Acc@1:  71.094 ( 82.265)  Acc@5:  98.438 ( 96.134)
Test: [ 380/391]  Time: 0.038s (0.358s,  357.96/s)  Loss:  0.8698 (0.9714)  Acc@1:  84.375 ( 82.226)  Acc@5:  96.875 ( 96.120)
Test: [ 390/391]  Time: 0.126s (0.358s,  223.47/s)  Loss:  1.4484 (0.9686)  Acc@1:  63.750 ( 82.262)  Acc@5:  93.750 ( 96.144)
 * Acc@1 82.262 (17.738) Acc@5 96.144 (3.856)
--result
{
    "model": "mamba_vision_T",
    "top1": 82.262,
    "top1_err": 17.738,
    "top5": 96.144,
    "top5_err": 3.856,
    "param_count": 31.79,
    "img_size": 224,
    "crop_pct": 1.0,
    "interpolation": "bicubic"
}
```

For reference, I have printed the following where these results are produced:

```
Python 3.10.13 [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch;print(torch.__version__)
2.1.1
>>> import timm; print(timm.__version__)
0.9.3dev0
>>> import einops;print(einops.__version__)
0.7.0
>>> import transformers;print(transformers.__version__)
4.37.1
```

We have rounded up the accuracy of 82.262 to 82.3 which is reasonable and common. 

I hope this addresses your question. 

Kind Regards
